src/eventstream-proxy/eventstream-proxy.base.ts

```diff
@@ -0,0 +1,85 @@
+import { Logger } from '@nestjs/common';
+import { SubscribeMessage } from '@nestjs/websockets';
+import { Event, EventStreamReply } from '../event-stream/event-stream.interfaces';
+import { EventStreamService, EventStreamSocket } from '../event-stream/event-stream.service';
+import { WebSocketEventsBase, WebSocketEx } from '../websocket-events/websocket-events.base';
+
+/**
+ * Base class for a websocket gateway that listens for and proxies event stream messages.
+ *
+ * To create the actual gateway, subclass and decorate your child, e.g.:
+ * @WebSocketGateway({ path: '/api/stream' })
+ */
+export abstract class EventStreamProxyBase extends WebSocketEventsBase {
+  socket?: EventStreamSocket;
+  url?: string;
+  topic?: string;
+
+  private awaitingAck = 0;
+
+  constructor(
+    protected readonly logger: Logger,
+    protected eventstream: EventStreamService,
+    requireAuth = false,
+  ) {
+    super(logger, requireAuth);
+  }
+
+  configure(url?: string, topic?: string) {
+    this.url = url;
+    this.topic = topic;
+  }
+
+  handleConnection(client: WebSocketEx) {
+    super.handleConnection(client);
+    if (this.server.clients.size === 1 && this.url !== undefined && this.topic !== undefined) {
+      this.logger.log(`Initializing event stream proxy`);
+      this.socket = this.eventstream.subscribe(
+        this.url,
+        this.topic,
+        events => {
+          this.awaitingAck += events.length;
+          for (const event of events) {
+            this.logger.log(`Proxying event: ${JSON.stringify(event)}`);
+            this.handleEvent(event);
+          }
+        },
+        receipt => {
+          this.handleReceipt(receipt);
+        },
+      );
+    }
+  }
+
+  handleDisconnect(client: WebSocketEx) {
+    super.handleDisconnect(client);
+    if (this.server.clients.size === 0) {
+      this.socket?.close();
+      this.socket = undefined;
+    }
+  }
+
+  protected handleEvent(_event: Event) {
+    // do nothing (can be overridden)
+  }
+
+  protected handleReceipt(_receipt: EventStreamReply) {
+    // do nothing (can be overridden)
+  }
+
+  ack() {
+    if (this.awaitingAck > 0) {
+      this.awaitingAck--;
```

Comment:
Suspect this is an area where we might see hard to debug issues, and a log entry (maybe debug level?) might be helpful for someone in the future.

Reply:
Yea, can add a log at least. Honestly it feels like this is not entirely the right solution. I wondered if I should either a) disable batching entirely by forcing batch size = 1, or b) propagate the batching back up to the client (would probably mean replacing the separate websocket events with a new "batch" event).

Reply:
This is overhauled in the latest changes.